File: Pinball App 2/Pinball App 2/settings/SettingsHomeLibrarySupport.swift
import SwiftUI

struct SettingsHomeLibrarySection: View {
    let manufacturers: [PinballCatalogManufacturerOption]
    let importedSources: [PinballImportedSourceRecord]
    let isEnabled: (String) -> Bool
    let isPinned: (String) -> Bool
    let onAddManufacturer: () -> Void
    let onAddVenue: () -> Void
    let onAddTournament: () -> Void
    let onToggleEnabled: (String, Bool) -> Void
    let onTogglePinned: (String, Bool) -> Void
    let onRefresh: (PinballImportedSourceRecord) -> Void
    let onDelete: (String) -> Void

    private var managedSources: [SettingsManagedSourceItem] {
        importedSources.map { source in
            SettingsManagedSourceItem(
                source: source,
                title: source.name,
                subtitle: managedSourceSubtitle(for: source),
                sourceType: source.type
            )
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            AppSectionTitle(text: "Library")

            AppCardSubheading(text: "Add")

            SettingsLibraryAddButtons(
                onAddManufacturer: onAddManufacturer,
                onAddVenue: onAddVenue,
                onAddTournament: onAddTournament
            )

            Text("Enabled adds that source's games to Library and Practice. Library adds the source to the Library source filter for quick switching. Up to \(PinballLibrarySourceStateStore.maxPinnedSources) sources can appear in Library at once.")
                .font(.caption)
                .foregroundStyle(.secondary)

            SettingsManagedSourceTable(
                items: managedSources,
                isEnabled: isEnabled,
                isPinned: isPinned,
                onToggleEnabled: onToggleEnabled,
                onTogglePinned: onTogglePinned,
                onRefresh: onRefresh,
                onDelete: onDelete
            )

            if importedSources.isEmpty {
                AppPanelEmptyCard(text: "No sources added yet.")
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(12)
        .appPanelStyle()
    }

    private func managedSourceSubtitle(for source: PinballImportedSourceRecord) -> String {
        switch source.type {
        case .manufacturer:
            return "Manufacturer • \(gameCountLabel(manufacturers.first(where: { $0.id == source.providerSourceID })?.gameCount ?? 0))"
        case .venue:
            return "Imported venue • \(gameCountLabel(source.machineIDs.count))"
        case .tournament:
            return "Match Play tournament • \(gameCountLabel(source.machineIDs.count))"
        case .category:
            return "Category"
        }
    }

    private func gameCountLabel(_ count: Int) -> String {
        count == 1 ? "1 game" : "\(count) games"
    }
}

private struct SettingsManagedSourceItem: Identifiable {
    let source: PinballImportedSourceRecord
    let title: String
    let subtitle: String
    let sourceType: PinballLibrarySourceType

    var id: String { source.id }
}

private struct SettingsLibraryAddButtons: View {
    let onAddManufacturer: () -> Void
    let onAddVenue: () -> Void
    let onAddTournament: () -> Void

    var body: some View {
        GeometryReader { proxy in
            let spacing: CGFloat = 8
            let totalWidth = max(0, proxy.size.width - (spacing * 2))
            let manufacturerUnits: CGFloat = 12
            let venueUnits: CGFloat = 5
            let tournamentUnits: CGFloat = 10
            let totalUnits = manufacturerUnits + venueUnits + tournamentUnits
            let manufacturerWidth = totalWidth * (manufacturerUnits / totalUnits)
            let venueWidth = totalWidth * (venueUnits / totalUnits)
            let tournamentWidth = totalWidth * (tournamentUnits / totalUnits)

            HStack(spacing: spacing) {
                Button("Manufacturer", action: onAddManufacturer)
                    .buttonStyle(AppCompactSecondaryActionButtonStyle(fillsWidth: true))
                    .frame(width: manufacturerWidth)

                Button("Venue", action: onAddVenue)
                    .buttonStyle(AppCompactSecondaryActionButtonStyle(fillsWidth: true))
                    .frame(width: venueWidth)

                Button("Tournament", action: onAddTournament)
                    .buttonStyle(AppCompactSecondaryActionButtonStyle(fillsWidth: true))
                    .frame(width: tournamentWidth)
            }
        }
        .frame(height: 34)
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

private struct SettingsManagedSourceTable: View {
    let items: [SettingsManagedSourceItem]
    let isEnabled: (String) -> Bool
    let isPinned: (String) -> Bool
    let onToggleEnabled: (String, Bool) -> Void
    let onTogglePinned: (String, Bool) -> Void
    let onRefresh: (PinballImportedSourceRecord) -> Void
    let onDelete: (String) -> Void
    @State private var pendingDeleteItem: SettingsManagedSourceItem?

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 12) {
                Text("Source")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(.secondary)
                    .frame(maxWidth: .infinity, alignment: .leading)
                Text("Enabled")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(.secondary)
                    .frame(width: 68)
                Text("Library")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(.secondary)
                    .frame(width: 68)
            }
            .padding(.bottom, 6)

            AppTableHeaderDivider()

            ForEach(Array(items.enumerated()), id: \.element.id) { index, item in
                SettingsManagedSourceRow(
                    item: item,
                    isEnabled: isEnabled(item.id),
                    isPinned: isPinned(item.id),
                    onToggleEnabled: { onToggleEnabled(item.id, $0) },
                    onTogglePinned: { onTogglePinned(item.id, $0) },
                    onRefresh: { onRefresh(item.source) },
                    onDelete: { pendingDeleteItem = item }
                )
                if index < items.count - 1 {
                    AppTableRowDivider()
                }
            }
        }
        .alert(deleteAlertTitle, isPresented: pendingDeleteItemAlertIsPresented) {
            Button("Delete", role: .destructive) {
                guard let pendingDeleteItem else { return }
                onDelete(pendingDeleteItem.id)
                self.pendingDeleteItem = nil
            }
            Button("Cancel", role: .cancel) {
                pendingDeleteItem = nil
            }
        } message: {
            Text(deleteAlertMessage)
        }
    }

    private var pendingDeleteItemAlertIsPresented: Binding<Bool> {
        Binding(
            get: { pendingDeleteItem != nil },
            set: { isPresented in
                if !isPresented {
                    pendingDeleteItem = nil
                }
            }
        )
    }

    private var deleteAlertTitle: String {
        let typeLabel = pendingDeleteItem.map(deleteTypeLabel(for:)) ?? "Source"
        return "Delete \(typeLabel)?"
    }

    private var deleteAlertMessage: String {
        guard let pendingDeleteItem else {
            return "This removes the source from Library and Practice."
        }
        return "Remove \(pendingDeleteItem.title) from Library and Practice? This cannot be undone."
    }

    private func deleteTypeLabel(for item: SettingsManagedSourceItem) -> String {
        switch item.sourceType {
        case .manufacturer:
            return "Manufacturer"
        case .venue:
            return "Venue"
        case .tournament:
            return "Tournament"
        case .category:
            return "Source"
        }
    }
}

private struct SettingsManagedSourceRow: View {
    let item: SettingsManagedSourceItem
    let isEnabled: Bool
    let isPinned: Bool
    let onToggleEnabled: (Bool) -> Void
    let onTogglePinned: (Bool) -> Void
    let onRefresh: () -> Void
    let onDelete: () -> Void

    private var canRefresh: Bool {
        item.sourceType == .venue || item.sourceType == .tournament
    }

    var body: some View {
        HStack(alignment: .center, spacing: 12) {
            VStack(alignment: .leading, spacing: 4) {
                AppCardSubheading(text: item.title)
                    .lineLimit(1)
                    .truncationMode(.tail)

                Text(item.subtitle)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .lineLimit(1)

                HStack(spacing: 6) {
                    if canRefresh {
                        Button("Refresh", action: onRefresh)
                            .buttonStyle(AppInlineActionChipButtonStyle())
                            .accessibilityLabel("Refresh \(item.title)")
                    }

                    Button("Delete", role: .destructive, action: onDelete)
                        .buttonStyle(AppInlineActionChipButtonStyle(isDestructive: true))
                        .accessibilityLabel("Remove \(item.title)")
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            Toggle("", isOn: Binding(get: { isEnabled }, set: onToggleEnabled))
                .labelsHidden()
                .frame(width: 68)
                .accessibilityLabel("Enabled for \(item.title)")

            Toggle("", isOn: Binding(get: { isPinned }, set: onTogglePinned))
                .labelsHidden()
                .frame(width: 68)
                .accessibilityLabel("Show \(item.title) in Library")
        }
        .padding(.vertical, 8)
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/settings/SettingsHomeLibrarySupport.kt
package com.pillyliu.pinprofandroid.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.pillyliu.pinprofandroid.library.CatalogManufacturerOption
import com.pillyliu.pinprofandroid.library.ImportedSourceRecord
import com.pillyliu.pinprofandroid.library.LibrarySourceState
import com.pillyliu.pinprofandroid.library.LibrarySourceStateStore
import com.pillyliu.pinprofandroid.library.LibrarySourceType
import com.pillyliu.pinprofandroid.ui.AppInlineActionChip
import com.pillyliu.pinprofandroid.ui.AppCardSubheading
import com.pillyliu.pinprofandroid.ui.AppConfirmDialog
import com.pillyliu.pinprofandroid.ui.AppPanelEmptyCard
import com.pillyliu.pinprofandroid.ui.AppSecondaryButton
import com.pillyliu.pinprofandroid.ui.AppSwitch
import com.pillyliu.pinprofandroid.ui.CardContainer
import com.pillyliu.pinprofandroid.ui.SectionTitle

@Composable
internal fun SettingsLibrarySection(
    manufacturers: List<CatalogManufacturerOption>,
    importedSources: List<ImportedSourceRecord>,
    sourceState: LibrarySourceState,
    onOpenAddManufacturer: () -> Unit,
    onOpenAddVenue: () -> Unit,
    onOpenAddTournament: () -> Unit,
    onToggleEnabled: (String, Boolean) -> Unit,
    onTogglePinned: (String, Boolean) -> Unit,
    onRefreshSource: (ImportedSourceRecord) -> Unit,
    onDeleteSource: (String) -> Unit,
) {
    var pendingDeleteSource by remember { mutableStateOf<ImportedSourceRecord?>(null) }

    CardContainer {
        SectionTitle("Library")
        AppCardSubheading("Add")
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            AppSecondaryButton(
                onClick = onOpenAddManufacturer,
                modifier = Modifier.weight(12f),
                contentPadding = PaddingValues(horizontal = 10.dp, vertical = 0.dp),
            ) { Text("Manufacturer", maxLines = 1, overflow = TextOverflow.Clip) }
            AppSecondaryButton(
                onClick = onOpenAddVenue,
                modifier = Modifier.weight(5f),
                contentPadding = PaddingValues(horizontal = 10.dp, vertical = 0.dp),
            ) { Text("Venue", maxLines = 1, overflow = TextOverflow.Clip) }
            AppSecondaryButton(
                onClick = onOpenAddTournament,
                modifier = Modifier.weight(10f),
                contentPadding = PaddingValues(horizontal = 10.dp, vertical = 0.dp),
            ) { Text("Tournament", maxLines = 1, overflow = TextOverflow.Clip) }
        }
        Text(
            "Enabled adds that source's games to Library and Practice. Library adds the source to the Library source filter for quick switching. Up to ${LibrarySourceStateStore.MAX_PINNED_SOURCES} sources can appear in Library at once.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        importedSources.forEach { source ->
            ManagedSourceRow(
                title = source.name,
                subtitle = importedSourceSubtitle(source, manufacturers),
                enabled = sourceState.enabledSourceIds.contains(source.id),
                pinned = sourceState.pinnedSourceIds.contains(source.id),
                canDisable = true,
                onEnabledChange = { isEnabled -> onToggleEnabled(source.id, isEnabled) },
                onPinnedChange = { isPinned -> onTogglePinned(source.id, isPinned) },
                onRefresh = if (source.type == LibrarySourceType.VENUE || source.type == LibrarySourceType.TOURNAMENT) {
                    { onRefreshSource(source) }
                } else {
                    null
                },
                onDelete = { pendingDeleteSource = source },
            )
        }
        if (importedSources.isEmpty()) {
            AppPanelEmptyCard(text = "No sources added yet.")
        }
    }

    pendingDeleteSource?.let { source ->
        AppConfirmDialog(
            title = "Delete ${source.type.deleteLabel()}?",
            message = "Remove ${source.name} from Library and Practice? This can't be undone.",
            confirmLabel = "Delete",
            onConfirm = {
                onDeleteSource(source.id)
                pendingDeleteSource = null
            },
            onDismiss = { pendingDeleteSource = null },
        )
    }
}

@Composable
private fun ManagedSourceRow(
    title: String,
    subtitle: String,
    enabled: Boolean,
    pinned: Boolean,
    canDisable: Boolean,
    onEnabledChange: (Boolean) -> Unit,
    onPinnedChange: (Boolean) -> Unit,
    onRefresh: (() -> Unit)?,
    onDelete: (() -> Unit)?,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            AppCardSubheading(title)
            Text(subtitle, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                onRefresh?.let { AppInlineActionChip(text = "Refresh", onClick = it) }
                onDelete?.let { AppInlineActionChip(text = "Delete", onClick = it, destructive = true) }
            }
        }
        Column(horizontalAlignment = Alignment.End) {
            Text("Enabled", style = MaterialTheme.typography.labelSmall)
            AppSwitch(
                checked = enabled,
                onCheckedChange = onEnabledChange,
                enabled = canDisable,
            )
        }
        Column(horizontalAlignment = Alignment.End) {
            Text("Pinned", style = MaterialTheme.typography.labelSmall)
            AppSwitch(
                checked = pinned,
                onCheckedChange = onPinnedChange,
            )
        }
    }
}

private fun LibrarySourceType.deleteLabel(): String = when (this) {
    LibrarySourceType.MANUFACTURER -> "Manufacturer"
    LibrarySourceType.VENUE -> "Venue"
    LibrarySourceType.TOURNAMENT -> "Tournament"
    LibrarySourceType.CATEGORY -> "Source"
}

private fun importedSourceSubtitle(
    source: ImportedSourceRecord,
    manufacturers: List<CatalogManufacturerOption>,
): String =
    when (source.type) {
        LibrarySourceType.MANUFACTURER -> {
            val count = manufacturers.firstOrNull { it.id == source.providerSourceId }?.gameCount ?: 0
            "Manufacturer • ${if (count == 1) "1 game" else "$count games"}"
        }

        LibrarySourceType.VENUE -> {
            val count = source.machineIds.size
            "Imported venue • ${if (count == 1) "1 game" else "$count games"}"
        }

        LibrarySourceType.TOURNAMENT -> {
            val count = source.machineIds.size
            "Match Play tournament • ${if (count == 1) "1 game" else "$count games"}"
        }

        LibrarySourceType.CATEGORY -> "Category"
    }
